File: android/app/src/main/kotlin/io/dotnetnativeinterop/transport/pb/AeadFrameCipher.kt
package io.dotnetnativeinterop.transport.pb

import java.nio.ByteBuffer
import java.nio.ByteOrder
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Per-frame AES-256-GCM for the framed-protobuf secure channel — the client mirror of the engine's
 * `AeadFrameCipher`. Each direction has its OWN key (HKDF with a distinct info label; see
 * [PqHandshakeClient]) and its OWN monotonic frame counter used as the GCM nonce, so a nonce is never
 * reused under a key.
 *
 * On the wire an encrypted frame payload is `ciphertext || tag(16)`; the 12-byte nonce is implicit (both
 * endpoints derive it from their frame counter) and never transmitted. `javax.crypto` AES/GCM/NoPadding
 * appends the 16-byte GCM tag to the ciphertext on encrypt — the exact `ciphertext || tag` layout the
 * engine's `AesGcm.Encrypt(cipherSpan, tagSpan)` produces — so the two are byte-compatible. No AAD is
 * passed (a deliberate match: there is no cleartext header to bind; the monotonic counter=nonce already
 * authenticates frame order).
 */
internal class AeadFrameCipher(private val sendKey: ByteArray, private val recvKey: ByteArray) {
    private var sendCounter = 0L
    private var recvCounter = 0L

    /** Encrypts one frame payload; returns `ciphertext || tag`. Advances the send counter. */
    fun encryptOutbound(plaintext: ByteArray): ByteArray {
        val nonce = nonceFor(sendCounter++)
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(sendKey, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        return cipher.doFinal(plaintext)
    }

    /**
     * Decrypts one `ciphertext || tag` frame payload. Advances the receive counter. Throws
     * `javax.crypto.AEADBadTagException` (the client-side analogue of the engine's
     * `AuthenticationTagMismatchException`) if the frame was tampered with, forged, or reordered.
     */
    fun decryptInbound(frame: ByteArray): ByteArray {
        val nonce = nonceFor(recvCounter++)
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(recvKey, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        return cipher.doFinal(frame)
    }

    companion object {
        /** Cipher name for the Trust inspector's live-params display (matches the engine's label). */
        const val ALGORITHM = "AES-256-GCM"

        private const val TRANSFORMATION = "AES/GCM/NoPadding"
        private const val TAG_BITS = 128   // 16-byte GCM tag (also CryptoKit's required size on Apple)
        private const val NONCE_BYTES = 12

        // Nonce = little-endian frame counter in the low 8 bytes, high 4 bytes zero — identical to the
        // engine's BinaryPrimitives.WriteInt64LittleEndian(nonce, counter) into a 12-byte buffer.
        private fun nonceFor(counter: Long): ByteArray {
            val nonce = ByteArray(NONCE_BYTES)
            ByteBuffer.wrap(nonce).order(ByteOrder.LITTLE_ENDIAN).putLong(counter)
            return nonce
        }
    }
}

File: ios/Shared/Pb/AeadFrameCipher.swift
import CryptoKit
import Foundation

/// Per-frame AES-256-GCM for the framed-protobuf secure channel — the Swift/CryptoKit mirror of the
/// engine's `AeadFrameCipher` and the Android one. Each direction has its OWN key (HKDF with a distinct
/// info label; see `PqHandshakeClient`) and its OWN monotonic frame counter used as the GCM nonce, so a
/// nonce is never reused under a key.
///
/// On the wire an encrypted frame payload is `ciphertext || tag(16)`; the 12-byte nonce is implicit (both
/// endpoints derive it from their frame counter) and never transmitted. CryptoKit's `SealedBox` exposes
/// `.ciphertext` and `.tag` separately — concatenated they are the exact `ciphertext || tag` layout the
/// engine's `AesGcm.Encrypt(cipherSpan, tagSpan)` and Android's `javax.crypto` produce. No AAD (a
/// deliberate match: the monotonic counter=nonce already authenticates frame order).
///
/// CryptoKit's ML-KEM/ML-DSA + this class's `AES.GCM` explicit-nonce API are iOS 26+. The app targets
/// iOS 17, so the whole PQ path is gated `@available(iOS 26.0, *)`; on older OS the binary transport
/// stays plaintext (honest disclosure in the Trust inspector — the same policy as FoundationModels).
@available(iOS 26.0, macOS 15.0, *)
final class AeadFrameCipher {
    private let sendKey: SymmetricKey
    private let recvKey: SymmetricKey
    private var sendCounter: UInt64 = 0
    private var recvCounter: UInt64 = 0

    static let algorithm = "AES-256-GCM"
    private static let tagBytes = 16

    init(sendKey: SymmetricKey, recvKey: SymmetricKey) {
        self.sendKey = sendKey
        self.recvKey = recvKey
    }

    /// Encrypts one frame payload; returns `ciphertext || tag`. Advances the send counter.
    func encryptOutbound(_ plaintext: Data) throws -> Data {
        let nonce = try Self.nonce(for: sendCounter)
        sendCounter += 1
        let sealed = try AES.GCM.seal(plaintext, using: sendKey, nonce: nonce)
        return sealed.ciphertext + sealed.tag   // nonce is implicit; never on the wire
    }

    /// Decrypts one `ciphertext || tag` frame payload. Advances the receive counter. Throws
    /// `CryptoKitError.authenticationFailure` (the analogue of the engine's tag-mismatch) on a tampered,
    /// forged, or reordered frame.
    func decryptInbound(_ frame: Data) throws -> Data {
        guard frame.count >= Self.tagBytes else { throw PbTransportError.closedEarly }
        let nonce = try Self.nonce(for: recvCounter)
        recvCounter += 1
        let split = frame.count - Self.tagBytes
        let ciphertext = Data(frame.prefix(split))     // fresh Data — normalize slice indices to 0
        let tag = Data(frame.suffix(Self.tagBytes))
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        return try AES.GCM.open(box, using: recvKey)
    }

    /// Nonce = little-endian frame counter in the low 8 bytes, high 4 bytes zero — identical to the
    /// engine's `BinaryPrimitives.WriteInt64LittleEndian` into a 12-byte buffer and Android's `nonceFor`.
    private static func nonce(for counter: UInt64) throws -> AES.GCM.Nonce {
        var bytes = [UInt8](repeating: 0, count: 12)
        withUnsafeBytes(of: counter.littleEndian) { src in
            for i in 0..<8 { bytes[i] = src[i] }
        }
        return try AES.GCM.Nonce(data: Data(bytes))
    }
}
